### DISCUSSION TITLE
2023/10/24/functional-domain-modeling-elixir/

### DISCUSSION THREAD

**Discussion by giscus** [katafrakt/website_comments#1] [Announcements]
# 2023/10/24/functional-domain-modeling-elixir/

How to apply functional domain modeling techniques to make your Elixir code cleaner, easier to follow and more test-friendly

https://katafrakt.me/2023/10/24/functional-domain-modeling-elixir/?utm_medium=email&utm_source=elixir-radar

**Comment by ktec** [comment]
Great post! I definitely agree with this approach, thanks for sharing.

I think you could use the "with" in your final version:


with ok <- Domain.borrow_book(subscription, currently_borrowed, unpaid_fines) do
      Borrowing.create_changeset(patron_id, book_id, DateTime.utc_now())
      |> Repo.insert()
 end


Because the it will automatically return any errors.

**Reply by katafrakt** [reply]
Thank you for the kind words!
You are right that it could be written with `with`. I'm just too used to Credo disallowing `with` with a single statement (which is enforced where I work) and wrote that using `case` without much thinking.

**Comment by hackvan** [comment]
Great post, its a nice way to separate concerns 👌🏼

A possible way to refactor the BorrowBook domain could be using pattern match, something like:

```elixir
def borrow_book(nil, _currently_borrowed, _unpaid_fines), do:
  {:error, :no_active_subscription}

def borrow_book(_active_subscription, currently_borrowed, _unpaid_fines)
  when length(currently_borrowed) >= 5, do: {:error, :too_many_borrowed_books}

def borrow_book(_active_subscription, _currently_borrowed, unpaid_fines)
  when length(unpaid_fines) > 0, do: {:error, :unpaid_fines}

def borrow_book(_active_subscription, _currently_borrowed, _unpaid_fines), do:
  :ok
```

**Comment by tomdonarski** [comment]
If I'd say I wouldn't welcome the follow-up post about how to tackle the explosion of the arguments passed to the domain function with open arms - I'd lie. :grin:

Not sure if that'd be a valid suggestion but judging by the title I doubted it'd benefit a ruby-guy like me.
So maybe something like 'Functional domain modeling (with Elixir examples)' would attract (or rather decrease the risk of putting-off non-elixir folks) more people.
Happily, it turned out I was wrong, and you've explained all the concepts in a neat and abstract way that applies generally :clap:.

I'll be certainly revisiting this write up numerous times.
